src/main/java/org/jabref/logic/openoffice/backend/Backend52.java

```diff
@@ -0,0 +1,465 @@
+        WrappedTargetException,
+        NoDocumentException {
+
+        Optional<Codec52.ParsedMarkName> op = Codec52.parseMarkName(refMarkName);
+        if (op.isEmpty()) {
+            throw new IllegalArgumentException("readCitationGroupFromDocumentOrThrow:"
+                                               + " found unparsable referenceMarkName");
+        }
+        Codec52.ParsedMarkName ov = op.get();
+        CitationGroupId cgid = new CitationGroupId(refMarkName);
+        List<Citation> citations = (ov.citationKeys.stream()
+                                    .map(Citation::new)
+                                    .collect(Collectors.toList()));
+
+        Optional<OOText> pageInfo =
+            (UnoUserDefinedProperty.getStringValue(doc, refMarkName)
+             .map(OOText::fromString));
+        pageInfo = PageInfo.normalizePageInfo(pageInfo);
+
+        setPageInfoInDataInitial(citations, pageInfo);
+
+        Optional<NamedRange> namedRange = citationStorageManager.nrmGetFromDocument(doc, refMarkName);
+
+        if (namedRange.isEmpty()) {
+            throw new IllegalArgumentException("readCitationGroupFromDocumentOrThrow:"
+                                               + " referenceMarkName is not in the document");
+        }
+
+        CitationGroup cg = new CitationGroup(OODataModel.JabRef52,
+                                             cgid,
+                                             ov.citationType,
+                                             citations,
+                                             Optional.of(refMarkName));
+        this.cgidToNamedRange.put(cgid, namedRange.get());
+        return cg;
+    }
+
+    /**
+     *  Create a reference mark with the given name, at the end of position.
+     *
+     *  On return {@code position} is collapsed, and is after the inserted space, or at the end of
+     *  the reference mark.
+     *
+     *  @param position Collapsed to its end.
+     *  @param insertSpaceAfter We insert a space after the mark, that carries on format of
+     *                          characters from the original position.
+     */
+    public CitationGroup createCitationGroup(XTextDocument doc,
+                                             List<String> citationKeys,
+                                             List<Optional<OOText>> pageInfos,
+                                             CitationType citationType,
+                                             XTextCursor position,
+                                             boolean insertSpaceAfter)
+        throws
+        CreationException,
+        NoDocumentException,
+        WrappedTargetException,
+        NotRemoveableException,
+        PropertyExistException,
+        PropertyVetoException,
+        IllegalTypeException {
+
+        Objects.requireNonNull(pageInfos);
+        if (pageInfos.size() != citationKeys.size()) {
+            throw new IllegalArgumentException("pageInfos.size != citationKeys.size");
```

Comment:
I am not sure what to write here. The test and the message is not targeted to the end user,
but the programmer. Removed the message.

Reply:
But the message is shown to the end user in case an exception occurs. Maybe something like "File is corrupt. Unable to parse." or something. Remember that exceptions may only be used for really exceptional conditions in the program, not as a substitute to control the program flow. https://github.com/HugoMatilla/Effective-JAVA-Summary#57-use-exceptions-only-for-exceptional-conditions

Reply:
I would write `assert pageInfos.size() == citationKeys.size()` if assertions were on.

The test is only needed because the function receives two lists and I could not express
statically that they should have the same number of elements. The concern is not
user input or files, just the caller providing data in the expected shape.
Might be an overkill, since the only caller `Frontend.createCitationGroup`,
as well as the only caller of `Frontend.createCitationGroup`: `UpdateCitationMarkers.createAndFillCitationGroup` already contain the same test.

I was thinking about changing the parameters of the function to contain 
`List<Pair<String,Optional<OOText>>> citationKeysAndPageInfos` or maybe
`List<Citation>` to clarify that here we expect an optional pageInfo for each key.

But I decided that the first may be too verbose (and vague), and the second may be confusing
because we would pass in Citation instances that have no representation in the document.
(Normally we read Citation and CitationGroup instances from the document,
or create them in the document here, along with creating the instance.)

The source of the data here is not the file. The only caller is `Frontend.createCitationGroup`,
which is called (only) from: `UpdateCitationMarkers.createAndFillCitationGroup`.

This latter is called from 3 locations:

- (1) `EditInsert.insertCitationGroup` which still receives a single pageInfo from the GUI which it embeds in a list
using `OODataModel.fakePageInfos(pageInfo, nEntries);`
-  (2) `EditMerge.mergeCitationGroups` which calls the backend (through Frontend): 
  `List<Optional<OOText>> pageInfos = fr.backend.combinePageInfos(cgs);` 
-  (3) `EditSeparate.separateCitations` which removes a citation group and creates a citation group for each citation